docs(dev/events): modernize / expand event system docs

joshtrichards opened the pull request:
### ☑️ Resolves

The existing Events chapter was showing its age and in need of a refactor.

It was a bit confusing in some spots, wasn't up-to-date with current best practices, and lacked coverage of the dispatcher.

This PR modernizes and expands the documentation for Nextcloud Events:

- adds emission/dispatch coverage
- clarifies listener registration / instantiation / handler execution lifecycle
- aligns docs/examples with current best practices - e.g.,
  - use the OCP Event Dispatcher and IBootstrap patterns
  - current PHP best practices
- adds additional code examples:
  - e.g., defining events, registering listeners, and emitting/dispatching events, usage of event payloads, and DI
- improves coverage of event payloads
- de-emphasize legacy approaches, referencing them only for migration purposes
  - e.g., hooks, public emitters, pre-IBootstrap listener patterns
- improves clarity throughout the documentation
  - e.g., correcting typos, clarifying ambiguous statements, and removing (most - *ahem*) informal phrasing
- new table of contents

These changes ensure app authors have clear, up-to-date, and actionable guidance for both consuming and producing events in Nextcloud apps.

### 🖼️ Screenshots

<img width="1080" height="32766" alt="image" src="https://github.com/user-attachments/assets/6e4b2bdf-063f-43e4-a05d-479a5cbdebc7" />

<!--
Please add a screenshot of your changed or added page(s).
This helps reviewers to quickly see how the resulting
lists, code blocks, headers and links look.
-->

joshtrichards committed:
docs(dev/events): modernize event system docs, add emission/dispatch examples, clarify listener lifecycle

Expanded the events documentation w/ detailed explanations, examples, and deprecated hooks and emitters sections.

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: fixup formatting

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: Fix typo

Signed-off-by: Josh <josh.t.richards@gmail.com>

come-nc commented on developer_manual/basics/events.rst:
```diff
@@ -4,166 +4,385 @@
+    /**
+     * Listener that adds two to a counter whenever an AddEvent is fired.
+     */
+    class AddTwoListener implements IEventListener {
```
```suggestion
     * @template-implements IEventListener<AddEvent>
     */
    class AddTwoListener implements IEventListener {
```

come-nc commented on developer_manual/basics/events.rst:
```diff
@@ -4,166 +4,385 @@
-A class that can handle an event will implement the ``\OCP\EventDispatcher\IEventListener`` interface. Class names should end with `Listener`.
+.. note::
+    By default there is no persistent "listener object" kept by the dispatcher. Each time the event is fired, the DI container will lazily instantiate a new instance of the listener (the class), invoke the handle() method, and then discard the instance. There are more advanced approaches to listener registration (singleton/shared), but
+    that is a more advanced use case -- not the default for DI-registered event listeners in Nextcloud. You may find this approach in core, but rarely in apps.
```
Is that true? I would expect core to use DI to get the listener, so as a singleton by default. Meaning if 2 instance of the same events are fired in the same request the same instance of the listener will be used.

joshtrichards committed:
chore(events): fixup dispatch() use versus intended dispatchTyped()

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): fix some typos/inaccuracies in event descriptions

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): fix some mistakes in examples

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
docs(events): fix inaccuracy in and clarify listener registration / instantiation

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): fixup per review comments

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
docs(events): add note about resolution via app container

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): fixup heading

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): clarity parent constructor compatibility shim

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
docs(dev): add missing events and additional clarifications

Also added new subsections for stopping, blocking, and broadcasting. And updated deprecated section: added GenericEvent and re-organized.

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore(events): fix links

Signed-off-by: Josh Richards <josh.t.richards@gmail.com>
